### PR TITLE
feat: `validateIf` for validation options

### PR DESCRIPTION
**Pull request by aoi-umi**
## Description

Add `validateIf` for validation options.
If you want to validate that condition by object, you can use `validation validateIf`.

`ValidateIf` will skip all other decorators, `validation validateIf` only effect the decorator that use this option

## Checklist

- [x] the pull request title describes what this PR does (not a vague title like `Update index.md`)
- [x] the pull request targets the *default* branch of the repository (`develop`)
- [x] the code follows the established code style of the repository
  - `npm run prettier:check` passes
  - `npm run lint:check` passes
- [x] tests are added for the changes I made (if any source code was modified)
- [x] documentation added or updated
- [x] I have run the project locally and verified that there are no errors

## Fixes
references https://github.com/typestack/class-validator/issues/1455

### CODE REVIEW

**Review by Franco1312 — DISMISSED** [review]

**Comment by braaar** [comment]
I think I prefer the `skip` or `skipIf` name for this option. I think `validateIf` is a bit confusing, since you can get it mixed up with the `ValidateIf` decorator.

**Comment by braaar** [comment]
Handy feature, though!

**Review by braaar — COMMENTED** [review]
I think this is a good proposal, but I think the function should be normalized to match the predicate function used in `ValidateIf`. 

This is how I see it being the most useful:
```ts
@Max(5, { validateIf: (o) => Boolean(o.someOtherProperty) })
```

I am now of the opinion that `validateIf` might be better than `skip` / `skipIf`.

**Comment by braaar** [comment]
See [my recent comment](https://github.com/typestack/class-validator/issues/1721#issuecomment-1857290706) in #1721

**Review by braaar — CHANGES_REQUESTED** [review]
As discussed in #1721, I think this decorator should have the same signature as `ValidateIf`. 

The signature `(value: any, validationArguments: ValidationArguments) => boolean;` makes it unnecessarily cumbersome to do simple stuff like: 
`@IsUUID({ validateIf: o => o.type === "something" })`

**Comment by aoi-umi** [comment]
updated @braaar

**Comment by kevit-kelvin-mandlik** [comment]
@braaar Please merge this feature I am waiting for so long...

**Comment by kevit-kelvin-mandlik** [comment]
@braaar @aoi-umi  What are you guys waiting upon this? Can we please move forward with this?

**Comment by aoi-umi** [comment]
> @braaar @aoi-umi What are you guys waiting upon this? Can we please move forward with this?

haha, i don't know either

**Comment by braaar** [comment]
We are discussing this feature internally right now. Hold on just a little longer, @kevit-kelvin-mandlik and @aoi-umi

**Comment by kevit-kelvin-mandlik** [comment]
@braaar Are we good to go on this??

**Comment by yasaichi** [comment]
Any updates on this PR? I've been looking forward to using this feature for a quite long time!

**Comment by braaar** [comment]
Hi again. Could I ask that you rewrite the docs and tests so that they more clearly illustrate the usefulness of this validation option? Personally I find the example in this PR a bit hard to parse. What does this actually mean in practice, you know?

```js
      validateIf: (obj: MyClass, value) => {
        return !obj.someOtherProperty || obj.someOtherProperty === 'min';
      },
```

I feel like that makes a poor case for this validation option. Adding a new feature should make people go "Nice! I like this!" rather than "Huh?"

**Comment by kelvinmandlik** [comment]
@aoi-umi @braaar Can we please please move forward with this it's been so long for one small feature.

**Review by braaar — DISMISSED** [review]
Let's ship this!

**Comment by braaar** [comment]
I just noticed that codecov is picking up a line that is untested. Could you see to this, @aoi-umi?

**Comment by aoi-umi** [comment]
> I just noticed that codecov is picking up a line that is untested. Could you see to this, @aoi-umi? 

haha, i almost forget this.
the case `'should validate none'` should cover, it seems cause by source map
image
image

**Comment by vladyslav-vin** [comment]
Hello, any updates or blockers? Been waiting for this feature!

**Comment by aoi-umi** [comment]
@braaar please check

**Comment by MaxKoldun** [comment]
@braaar Please merge this feature. I need it for my project.

**Comment by braaar** [comment]
@aoi-umi check the latest CI / Linters action. The codecov is still not happy with the sourceMap change. Could you try refactoring the code in some way (perhaps inverting or re-ordering some if-statements) to see if that makes codecov happy?

**Comment by aoi-umi** [comment]
sorry i forget to run the lint, it is ok now? @braaar

**Comment by braaar** [comment]
Thanks. I'm still hesitant to merge due to the codecov problem. I'm not a test coverage hawk, I'm just worried that it might be catching something we aren't catching.

**Comment by braaar** [comment]
I think it's time.

**Review by braaar — APPROVED** [review]
